Futura/Futura/Stream/Signal.swift

```diff
@@ -0,0 +1,135 @@
+/* Copyright 2018 Miquido
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License. */
+
+public class Signal<Value> {
+    internal typealias Token = Either<Error, Value>
+    internal typealias Subscriber = (Either<Error?, Token>) -> Void
+
+    private var subscriptionID: Subscription.ID = 0
+    private let privateCollector: SubscriptionCollector = .init()
+
+    internal let lock: RecursiveLock = .init()
+    internal var subscribers: [(id: Subscription.ID, subscriber: Subscriber)] = .init()
+    internal weak var collector: SubscriptionCollector?
+    internal var isFinished: Bool = false
+    internal var isUnsubscribing: Bool = false
+
+    internal init(collector: SubscriptionCollector?) {
+        self.collector = collector
+    }
+
+    internal func subscribe(_ subscriber: @escaping Subscriber) -> Subscription? {
+        return lock.synchronized {
+            guard !isFinished else { return nil }
+            let id = subscriptionID.next()
+            subscribers.append((id: id, subscriber: subscriber))
+            return Subscription.init { [weak self] in
+                guard let self = self else { return }
+                self.lock.synchronized {
+                    self.isUnsubscribing = true
+                    if let idx = self.subscribers.firstIndex(where: { $0.id == id }) {
+                        self.subscribers.remove(at: idx)
+                    } else { /* do nothing */ }
+                    self.isUnsubscribing = false
+                }
+            }
+        }
+    }
+
+    internal func broadcast(_ token: Token) {
+        lock.synchronized {
+            guard !isSuspended else { return }
+            subscribers.forEach { $0.1(.right(token)) }
```

Comment:
`$0.1(.right` hard to understand here :)

Reply:
yeah... I know, I am still wondering if keeping Result and making additional enum dedicated for Signals was better or not... I will give it one more chance and see if it looks better when Token is a dedicated enum